src/common/meta/src/reconciliation/utils.rs

```diff
@@ -433,6 +433,105 @@ pub(crate) async fn validate_table_id_and_name(
+pub(crate) fn resolve_column_metadatas_for_logical_table(
+    region_metadatas: &[RegionMetadata],
+    table_info: &RawTableInfo,
+) -> Result<Vec<ColumnMetadata>> {
+    let is_same_table = region_metadatas
+        .windows(2)
+        .all(|w| w[0].region_id.table_id() == w[1].region_id.table_id());
+
+    ensure!(
+        is_same_table,
+        UnexpectedSnafu {
+            err_msg: "Region metadatas are not from the same table"
+        }
+    );
+
+    let region_metadata = region_metadatas
+        .iter()
+        .max_by_key(|r| r.column_metadatas.len())
+        .unwrap();
+
+    let is_invariant_preserved = check_column_metadatas_invariants_for_logical_table(
+        &region_metadata.column_metadatas,
+        table_info,
+    );
+    ensure!(
+        is_invariant_preserved,
+        UnexpectedSnafu {
+            err_msg: format!(
+                "Column metadata invariants violated for region {}. Resolved column metadata: {:?}, region column metadata: {:?}",
+                region_metadata.region_id,
+                region_metadata.column_metadatas.iter().map(ColumnMetadataDisplay).collect::<Vec<_>>(),
+                region_metadata.column_metadatas.iter().map(ColumnMetadataDisplay).collect::<Vec<_>>()
+            )
+        }
+    );
+
+    Ok(region_metadata.column_metadatas.clone())
+}
+
+/// Checks whether the column metadata invariants hold for the logical table.
+///
+/// Invariants:
+/// - Primary key (Tag) columns must exist in the new metadata.
+/// - Timestamp column must remain exactly the same in name and ID.
+///
+/// TODO(weny): add tests
+pub(crate) fn check_column_metadatas_invariants_for_logical_table(
+    column_metadatas: &[ColumnMetadata],
+    table_info: &RawTableInfo,
+) -> bool {
+    let new_primary_keys = column_metadatas
+        .iter()
+        .filter(|c| c.semantic_type == SemanticType::Tag)
+        .map(|c| c.column_schema.name.as_str())
+        .collect::<HashSet<_>>();
+
+    let old_primary_keys = table_info
+        .meta
+        .primary_key_indices
+        .iter()
+        .map(|i| table_info.meta.schema.column_schemas[*i].name.as_str());
+
+    for name in old_primary_keys {
+        if !new_primary_keys.contains(name) {
+            return false;
+        }
+    }
+
+    let old_timestamp_column_name = table_info
+        .meta
+        .schema
+        .column_schemas
+        .iter()
+        .find(|c| c.is_time_index())
+        .map(|c| c.name.as_str());
+
+    let new_timestamp_column_name = column_metadatas
+        .iter()
+        .find(|c| c.semantic_type == SemanticType::Field)
+        .map(|c| c.column_schema.name.as_str());
+
+    old_timestamp_column_name != new_timestamp_column_name
+}
+
+/// Returns true if the logical table info needs to be updated.
+pub(crate) fn need_update_logical_table_info(
+    table_info: &RawTableInfo,
+    column_metadatas: &[ColumnMetadata],
+) -> bool {
+    table_info.meta.schema.column_schemas.len() != column_metadatas.len()
```

Comment:
is simply checking the sizes equality sufficient to determine the update or not? what if the sizes are same but the column schemas are not equal

Reply:
The logical table only supports adding columns; modifying column types or dropping columns is not supported.